packages/widget-markdown/src/extension.ts

```diff
@@ -0,0 +1,135 @@
+  // derive deterministic ID for files from path
+  const FILE_UUID_NAMESPACE = '1b671a64-40d5-491e-99b0-da01ff1f3341'
+  return uuid(path, FILE_UUID_NAMESPACE)
+}
+
+const uriToMarkdownDocument = (fileUri: vscode.Uri) => ({
+  id: getMarkdownDocumentId(fileUri.path),
+  path: fileUri.path,
+  name: path.basename(fileUri.path),
+  content: undefined,
+})
+
+const loadMarkdownDocuments = async () => {
+  const { workspace } = vscode
+  const fileUris = await workspace.findFiles('*.md')
+  const markdownDocuments = fileUris.map(uriToMarkdownDocument)
+
+  return markdownDocuments
+}
+
+export class MarkdownExtensionManager extends ExtensionManager<State, {}> {
+  constructor (context: vscode.ExtensionContext, channel: vscode.OutputChannel) {
+    super(
+      context,
+      channel,
+      STATE_KEY,
+      {},
+      { markdownDocuments: [], markdownDocumentSelected: undefined }
+    )
+
+    // Load initial documents
+    loadMarkdownDocuments().then((markdownDocuments) => {
+      this.updateState('markdownDocuments', markdownDocuments)
+      this.broadcast({ markdownDocuments })
+      if (markdownDocuments.length > 0) {
+        this.updateState('markdownDocumentSelected', markdownDocuments[0].id)
+        this.broadcast({ markdownDocumentSelected: markdownDocuments[0].id })
+      }
+    })
+
+    // keep watching for changes
+    const watcher = vscode.workspace.createFileSystemWatcher(
+      '**/*.md',
+      false,
+      true,
+      false
+    )
+    watcher.onDidCreate(this.addMarkdownDocument)
+    watcher.onDidDelete(this.removeMarkdownDocument)
+  }
+
+  loadMarkdownContent = async (doc: MarkdownDocument) => {
+    const uri = vscode.Uri.file(doc.path)
+    const content = (await vscode.workspace.fs.readFile(uri)).toString()
+    return { ...doc, content }
+  }
+
+  private addMarkdownDocument = async (uri: vscode.Uri) => {
+    const doc = await uriToMarkdownDocument(uri)
+    const oldMarkdownDocuments = this.state.markdownDocuments
+    const updatedMarkdownDocuments = [...oldMarkdownDocuments, doc]
+    this.updateState('markdownDocuments', updatedMarkdownDocuments)
+    this.broadcast({ markdownDocuments: updatedMarkdownDocuments })
+  }
+
+  private removeMarkdownDocument = (uri: vscode.Uri) => {
+    const id = getMarkdownDocumentId(uri.path)
+    const oldMarkdownDocuments = this.state.markdownDocuments
+    const updatedMarkdownDocuments = oldMarkdownDocuments.filter(
+      (doc) => doc.id !== id
+    )
+    this.updateState('markdownDocuments', updatedMarkdownDocuments)
+    this.broadcast({ markdownDocuments: updatedMarkdownDocuments })
+  }
+
+  updateMarkdownDocument = async (updatedDoc: MarkdownDocument) => {
+    const updatedMarkdownDocuments = this.state.markdownDocuments.map(
+      (currDoc) => {
+        if (currDoc.id === updatedDoc.id) {
+          return updatedDoc
+        }
+        return currDoc
+      }
+    )
+    this.updateState('markdownDocuments', updatedMarkdownDocuments)
```

Comment:
If I read this correctly we keep the content of all selected markdown documents in the state. While this is already better I think there is no need for it, particularly as documents change over time. I'ld suggest to remove the `content` property from `MarkdownDocument` and once a document got selected emit just the content, e.g. `tangle.emit('markdownDocumentContent', content)`.

So we basically have two distinct properties in our state: a list with all documents (including path and ids), and a string with the content of the current selected document.